lms/djangoapps/grades/models.py

```diff
@@ -311,74 +268,95 @@ def read_grade(cls, user_id, usage_key):
 
         Raises PersistentSubsectionGrade.DoesNotExist if applicable
         """
-        return cls.objects.get(
+        return cls.objects.select_related('visible_blocks').get(
             user_id=user_id,
             course_id=usage_key.course_key,  # course_id is included to take advantage of db indexes
             usage_key=usage_key,
         )
 
     @classmethod
-    def update_grade(
-            cls,
-            user_id,
-            usage_key,
-            course_version,
-            subtree_edited_timestamp,
-            earned_all,
-            possible_all,
-            earned_graded,
-            possible_graded,
-            visible_blocks,
-    ):
-        """
-        Updates a previously existing grade.
-
-        This is distinct from update() in that `grade.update()` operates on an
-        existing grade object, while this is a classmethod that pulls the grade
-        from the database, and then updates it.  If you already have a grade
-        object, use the update() method on that object to avoid an extra
-        round-trip to the database.  Use this classmethod if all you have are a
-        user and the usage key of an existing grade.
-
-        Requires all the arguments listed in docstring for create_grade
-        """
-        grade = cls.read_grade(
+    def bulk_read_grades(cls, user_id, course_key):
+        """
+        Reads all grades for the given user and course.
+
+        Arguments:
+            user_id: The user associated with the desired grades
+            course_key: The course identifier for the desired grades
+        """
+        return cls.objects.select_related('visible_blocks').filter(
             user_id=user_id,
-            usage_key=usage_key,
+            course_id=course_key,
         )
 
-        grade.update(
-            course_version=course_version,
-            subtree_edited_timestamp=subtree_edited_timestamp,
-            earned_all=earned_all,
-            possible_all=possible_all,
-            earned_graded=earned_graded,
-            possible_graded=possible_graded,
-            visible_blocks=visible_blocks,
+    @classmethod
+    def update_or_create_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
```

Comment:
Technically this is now a wrapper for the django-defined `cls.objects.update_or_create`, which is itself a wrapper for `cls.objects.update` or `cls.objects.create`.

Also, it looks like we no longer enumerate the list of parameters needed to create a model in any method.